src/ann_solo/spectrum.py

```diff
@@ -214,6 +215,57 @@ def spectrum_to_vector(spectrum: MsmsSpectrum, min_mz: float, max_mz: float,
     return vector
 
 
+def spectrum_to_vector(spectrum: MsmsSpectrum, transformation: ss.csr_matrix,
+                       min_mz: float, max_mz: float, bin_size: float, dim: int,
+                       norm: bool) -> np.ndarray:
+    """
+    Convert a single spectrum to a dense NumPy vector.
+
+    Peaks are first discretized to mass bins of width `bin_size` starting from
+    `min_mz`, after which they are transformed using sparse random projections.
+
+    Parameters
+    ----------
+    spectrum : MsmsSpectrum
+        The spectrum to be converted to a vector.
+    transformation : ss.csr_matrix
+        Sparse random projection transformation to convert sparse spectrum
+        vectors to low-dimensional dense vectors.
+    min_mz : float
+        The minimum m/z to include in the vector.
+    max_mz : float
+        The maximum m/z to include in the vector.
+    bin_size : float
+        The bin size in m/z used to divide the m/z range.
+    dim : int
+        The high-resolution vector dimensionality.
+    norm : bool
+        Normalize the vector to unit length or not.
+    Returns
+    -------
+    np.ndarray
+        The low-dimensional transformed spectrum vector with unit length.
+    """
+    # set the spectrum range between min and max mz
+    spectrum = spectrum.set_mz_range(min_mz, max_mz)
+    # Convert a spectrum to a binned sparse vector
+    data = np.array(spectrum.intensity, dtype=np.float32)
+    indices = np.array(
+        [math.floor((mz - min_mz) / bin_size) for mz in spectrum.mz],
+        dtype=np.int32)
+    indptr = np.array([0, len(spectrum.mz)], dtype=np.int32)
+
+    # Instantiate the sparse matrix
+    sparse_vector = ss.csr_matrix(
+        (data, indices, indptr), (1, dim), np.float32, False)
+
+    # Transform
+    transformed_vector = (sparse_vector @ transformation).toarray()
```

Comment:
**todo:** Maybe this step can be made more efficient when doing this on a batch of spectra with a single matrix multiplication? That should be vectorized in the backend, although it will require some refactoring of the upstream code.

Reply:
That is correct, but the whole logic has to change. In the whole code of ANN-SoLo, we are using generators to process the spectra as they are read. There will be lots of refactoring of the code and logic upstream which I think it is not the goal of this issue here. I would rather focus on assessing the benefits of this feature, and the neutral loss which will be straightforward to incorporate in the current solution,  rather than spending more time refactoring everything.

Can be added as an issue for refactoring, vectorizing, and testing of the whole new solution, and maybe switch to sparse random projection rather than gaussian random projection (which are usually dense). It might blow up the memory needed when loading the whole library for a matrix multiplication (to be tested).

Reply:
Yes, that makes sense.